aep/general/0146/aep.md.j2

```diff
@@ -1,5 +1,97 @@
+the field needs to be; in general, APIs **should** attempt to introduce the
+"least generic" approach that is able to satisfy the use case.
+
+For example, an API **should not** use a completely generic field (such as
+`google.protobuf.Struct` in protobuf APIs) when the value of the field must
+correspond to one of a known number of schemas. Instead, the API **should** use
+a `oneof` to represent the known schemas.
+
+### Generic fields in protobuf APIs
+
+#### Oneof
+
+A `oneof` **may** be used to introduce a type union: the user or API is able to
+specify one of the fields inside the `oneof`. Additionally, a `oneof` **may**
+be used with the same type (usually strings) to represent a semantic difference
+between the options.
+
+Because the individual fields in the `oneof` have different keys, a developer
+can programmatically determine which (if any) of the fields is populated.
+
+A `oneof` preserves the largest degree of type safety and semantic meaning for
+each option, and APIs **should** generally prefer them over other generic or
+polymorphic options when feasible. However, the `oneof` construct is ill-suited
+when there is a large (or unlimited) number of potential options, or when there
+is a large resource structure that would require a long series of "cascading
+oneofs".
+
+**Note:** Adding additional possible fields to an existing `oneof` is a
+non-breaking change, but moving existing fields into or out of a `oneof` is
+breaking (it creates a backwards-incompatible change in Go protobuf stubs).
+
+#### Maps
+
+Maps **may** be used in situations where many values _of the same type_ are
+needed, but the keys are unknown or user-determined.
+
+Maps are usually not appropriate for generic fields because the map values all
+share a type, but occasionally they are useful. In particular, a map can
+sometimes be suited to a situation where many objects of the same type are
+needed, with different behavior based on the names of their keys (for example,
+using keys as environment names).
+
+#### Struct
+
+The [`google.protobuf.Struct`][struct] object **may** be used to represent
+arbitrary nested JSON. Keys can be strings, and values can be floats, strings,
+booleans, arrays, or additional nested structs, allowing for an arbitrarily
+nested structure that can be represented as JSON (and is automatically
+represented as JSON when using REST/JSON).
+
+A `Struct` is most useful when the API does not know the schema in advance, or
+when a API needs to store and retrieve arbitrary but structured user data.
+Using a `Struct` is convenient for users in this case because they can easily
+get JSON objects that can be natively manipulated in their environment of
+choice.
+
+If a API needs to reason about the _schema_ of a `Struct`, it **should** use
+[JSONSchema][] for this purpose. Because JSONSchema is itself JSON, a valid
+JSONSchema document can itself be stored in a `Struct`.
```

Comment:
Why would you want to put JSONSchema on an "untyped" struct? You might as well just create regular proto messages

Reply:
agreed, I'm not super clear on why this is better than a proto struct.

There are interesting cases for representing really complex schemas in an untyped fashion though - for example, Cloud monitoring dashboards for GCP are extremely complex, and IIRC the GCP Terraform provider represents that as JSON.

Reply:
My interpretation of this guidance is if you need to support user-specified schemas, use JSON Schema for it.  Imagine that you're building a platform for user-generated games (hopefully not too much of a stretch), and you provide a generic key-value store for each game.  It might be useful to let your game creators tell you something about the schema of their data in order to support things like GDPR (which fields correspond to a user ID?), something like Django Admin or the Firestore GUI, something to facilitate in-game trading, etc.

The second sentence says this can be stored in a `Struct` - but a `Struct` alone doesn't provide schema semantics in the way that JSON Schema does.

Reply:
Discussed live and concluded this guidance is more or less fine as is.  Useful as an anti-bikeshedding tool when user-generated schemas have to be supported.